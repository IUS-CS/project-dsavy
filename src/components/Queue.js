import React, {useState} from 'react'
import {animated, useTransition} from 'react-spring'
import {Container, Row, Col} from 'react-bootstrap';
import Enqueue from './Forms/Enqueue'
import Dequeue from './Forms/Dequeue'

const Queue = () => {
  const [array, setArray] = useState([])
  const [index, setIndex] = useState(0)
  
  const queueStyle =  {
    position: 'absolute',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    background: "linear-gradient(#58C2ED, #1B85DC)",
    width: '5rem',
    height: '18rem',
    bottom: '-2.5px'
  }
  
  // updates state of array 
  const onAdd = (entry) => {
    setArray(enqueue(array, entry))
    setIndex(index + 1)
  }

  // returns array with number added to the end
  const enqueue = (queue, number) => {
    if (number == '') {
      alert('Enter number')
      return queue
    }
    
    if (queue.length > 9) {
      alert('Queue is at max length')
      return queue
    }

    if (Math.abs(number) > 999) {
      alert('Number must be 3 digits or less')
      return queue
    }
    
    queue.push({key: index, element: number})
    
    return queue
  }

  // updates state of array
  const onDelete = () => {
    setArray(dequeue(array))
    setIndex(index + 1)
  }

  // returns array with first element deleted
  const dequeue = (queue) => {
    return queue.splice(1)  
  }
  
  const transition = useTransition(array, entry => entry.key, {
    initial: {opacity: 0},
    from: {opacity: 0},
    enter: {opacity: 1, transition: 'opacity .25s'},
    leave: {opacity: 0, transition: 'opacity .25s', color: 'red'}
  })

  // display array with animation when added to or removed
  const displayQueue = () => {
    console.log("hello")
    return (transition.map(({item, props, key}) =>         
      <animated.div key={key} style={props}>
        <li style={{listStyle: 'none'}}>
          <Col style={Object.assign({border: 'solid', borderLeft: 0, left: `${array.indexOf(item)*5}rem`}, queueStyle)} >
            <h4 style={{color: 'white'}}>{item.element}</h4>
          </Col>
        </li>
      </animated.div>
    ))
  }

  return (
    <Container>
      <header className='header'>
        <h1>Queue</h1>
        <h1>FIFO: First In First Out</h1>
      </header>
      <Row><Enqueue enqueue={onAdd} /></Row>
      <Row><Dequeue dequeue={onDelete} /></Row>
      <Row>
        <div style={{position: 'relative', width: '50rem', height: '18rem', margin: '0 auto' , border: 'solid'}}>
          <ul>{displayQueue()}</ul>
        </div>
      </Row>
    </Container>
  )
}

export default Queue