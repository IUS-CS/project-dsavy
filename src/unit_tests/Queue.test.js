import React from 'react';
import Queue from 'components/Queue';
import renderer from 'react-test-renderer';

// Queue renders
test("Enqueue renders correctly", () => {
    const tree = renderer.create(<Queue/>).toJSON()
    expect(tree).toMatchSnapshot()
});

describe('Testing enqueue', () => {
    const enqueue = (queue, number) => {
        if (number == '') {
          return queue
        }
        
        if (queue.length > 9) {
          return queue
        }
    
        if (Math.abs(number) > 999) {
          return queue
        }
        
        queue.push(number)
        return queue
    }

    it('enqueue correctly', ()=>{
      let array = [1]
      expect(enqueue(array,4)).toStrictEqual([1,4]);
    })

    it('enqueue does not add, number too large', ()=>{
      let array = [1]
      expect(enqueue(array,1000)).toStrictEqual([1]);
    })
    
    it('adds to empty queue', ()=>{
      let array = []
      expect(enqueue(array,950)).toStrictEqual([950]);
    })

    it('Fails to add, queue at max length', ()=>{
      let array = [1,2,3,4,5,6,7,8,9,10]
      expect(enqueue(array,950)).toStrictEqual([1,2,3,4,5,6,7,8,9,10]);
    })

    it('Fails to add, no entry', ()=>{
        let array = [1,2,3,4,5,6,7,8]
        expect(enqueue(array,'')).toStrictEqual([1,2,3,4,5,6,7,8]);
      })
});

describe('Testing enqueue', () => {
    const dequeue = (queue) => {
        return queue.splice(1)  
    }

    it('enqueue correctly', ()=>{
      let array = [1]
      expect(dequeue(array)).toStrictEqual([]);
    })

    it('enqueue does not add, number too large', ()=>{
      let array = [1,2,3,4]
      expect(dequeue(array)).toStrictEqual([2,3,4]);
    })
    
    it('adds to empty queue', ()=>{
      let array = []
      expect(dequeue(array)).toStrictEqual([]);
    })
});