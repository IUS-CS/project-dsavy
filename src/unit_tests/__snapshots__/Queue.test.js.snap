// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enqueue renders correctly 1`] = `
<div
  className="container"
>
  <header
    className="header"
  >
    <h1>
      Queue
    </h1>
    <h1>
      FIFO: First In First Out
    </h1>
  </header>
  <div
    className="row"
  >
    <form
      onSubmit={[Function]}
    >
      <button
        className="my-1 btn btn-primary"
        disabled={false}
        type="submit"
      >
        Enqueue
      </button>
      <input
        className="enqueue"
        onChange={[Function]}
        type="number"
        value=""
      />
    </form>
  </div>
  <div
    className="row"
  >
    <button
      className="my-1 btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Dequeue
    </button>
  </div>
  <div
    className="row"
  >
    <div
      style={
        Object {
          "border": "solid",
          "height": "18rem",
          "margin": "0 auto",
          "position": "relative",
          "width": "50rem",
        }
      }
    >
      <ul />
    </div>
  </div>
</div>
`;
